src/main/scala/ch/epfl/bluebrain/nexus/kg/resources/Resources.scala

```diff
@@ -78,6 +75,68 @@ object Resources {
   def get[F[_]](id: ResId, tag: String)(implicit repo: Repo[F]): OptionT[F, Resource] =
     repo.get(id, tag)
 
+  /**
+    * Updates an existing resource.
+    *
+    * @param id              the id of the resource
+    * @param rev             the last known revision of the resource
+    * @param additionalTypes a collection of additional (asserted or inferred) types of the resource
+    * @param source          the new source representation in json-ld format
+    * @return either a rejection or the updated resource in the F context
+    */
+  def update[F[_]: Monad: Resolution](
+      id: ResId,
+      rev: Long,
+      additionalTypes: Set[AbsoluteIri],
+      source: Json
+  )(implicit repo: Repo[F], identity: Identity): EitherT[F, Rejection, Resource] =
+    // format: off
+    for {
+      resource <- get(id, rev).toRight(NotFound(id.ref))
+      value       <- materialize[F](id, source)
+      graph       = value.graph
+      resolved    <- resolveWithImports(resource.schema)
+      _           <- validate(resolved.schema, resolved.schemaImports, resolved.dataImports, graph)
+      types       = joinTypes(graph, additionalTypes)
+      updated     <- repo.update(id, rev, types, source)
+    } yield updated
+  // format: on
+
+  /**
+    * Deprecates an existing resource
+    *
+    * @param id  the id of the resource
+    * @param rev             the last known revision of the resource
+    * @return Some(resource) in the F context when found and None in the F context when not found
+    */
+  def deprecate[F[_]](id: ResId, rev: Long)(implicit repo: Repo[F],
+                                            identity: Identity): EitherT[F, Rejection, Resource] =
+    repo.deprecate(id, rev)
+
+  /**
+    * Tags a resource. This operation alias the provided ''targetRev'' with the  provided ''tag''
+    *
+    * @param id        the id of the resource
+    * @param rev       the last known revision of the resource
+    * @param targetRev the revision that is being aliased with the provided ''tag''
+    * @param tag       the tag of the alias for the provided ''rev''
+    * @return Some(resource) in the F context when found and None in the F context when not found
+    */
+  def tag[F[_]](id: ResId, rev: Long, targetRev: Long, tag: String)(
+      implicit repo: Repo[F],
+      identity: Identity): EitherT[F, Rejection, Resource] =
+    repo.tag(id, rev, targetRev, tag)
+
+  private def resolveWithImports[F[_]: Monad: Resolution](schema: Ref) =
```

Comment:
rename to `schemaContext` to match the result type and add an explicit return type